packages/@aws-cdk/aws-rds/lib/cluster-engine.ts

```diff
@@ -0,0 +1,318 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
+import * as core from '@aws-cdk/core';
+import { ClusterParameterGroup, IParameterGroup, ParameterGroup } from './parameter-group';
+import { ParameterGroupFamilyMapping } from './private/parameter-group-family-mapping';
+import { compare } from './private/version';
+
+/**
+ * The extra options passed to the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineBindOptions {
+  /**
+   * The role used for S3 importing.
+   *
+   * @default - none
+   */
+  readonly s3ImportRole?: iam.IRole;
+
+  /**
+   * The role used for S3 exporting.
+   *
+   *  @default - none
+   */
+  readonly s3ExportRole?: iam.IRole;
```

Comment:
These are only relevant for aurora. They should instead be modeled as properties of the aurora engine.

Reply:
Few things:

1. I don't think they're actually relevant only for aurora. Let me experiment with it and verify.
2. If that's true, then we should add validation for the other engines that fails if those have been provided. So they still should be passed here.
3. These are properties of the Cluster though. That is, you can have 2 Clusters that have the same engine, but import/export from/to different S3 Buckets, using different Roles. I don't think it makes sense for them to be on Engine.

Reply:
Confirmed this works for PostgreSQL. Docs: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraPostgreSQL.Migrating.html#USER_PostgreSQL.S3Import

Reply:
>> I don't think they're actually relevant only for aurora. Let me experiment with it and verify.
>
> Confirmed this works for PostgreSQL. Docs: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraPostgreSQL.Migrating.html#USER_PostgreSQL.S3Import

The heading on that article is 'Importing Amazon S3 Data into an **Aurora** PostgreSQL DB Cluster'. Did you try this on a non-Aurora PostgreSQL?

Can you confirm that the API will be approximately, as we discussed -

```ts
new rds.DatabaseCluster(this, 'Cluster', {
  engine: new rds.AuroraClusterEngine({
    s3ImportRole: ...,
    s3ExportRole: ...,
    // ...
  }),
});
```

If not, what will the new API be when customers want to pass in `s3ImportRole` and `s3ExportRole`

Reply:
> > > I don't think they're actually relevant only for aurora. Let me experiment with it and verify.
> > 
> > 
> > Confirmed this works for PostgreSQL. Docs: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraPostgreSQL.Migrating.html#USER_PostgreSQL.S3Import
> 
> The heading on that article is 'Importing Amazon S3 Data into an **Aurora** PostgreSQL DB Cluster'. Did you try this on a non-Aurora PostgreSQL?

There is no cluster engine that's a non-Aurora PostgreSQL.

Either way, these options (S3 import/export) make sense for Clusters with all 3 currently supported engine types, which was the original question.

Reply:
> Can you confirm that the API will be approximately, as we discussed -
> 
> ```ts
> new rds.DatabaseCluster(this, 'Cluster', {
>   engine: new rds.AuroraClusterEngine({
>     s3ImportRole: ...,
>     s3ExportRole: ...,
>     // ...
>   }),
> });
> ```
> 
> If not, what will the new API be when customers want to pass in `s3ImportRole` and `s3ExportRole`

No. The API will be exactly as it is now - these properties can be set on a Cluster when creating it, which is the correct modeling (those are Cluster properties, not Engine properties(.

Reply:
Ok, we're back to square one then. 

When we discussed the design offline around #8506, we said that this was possible - 

```ts
new rds.DatabaseCluster(this, 'Cluster', {
  engine: rds.DatabaseClusterEngine.auroraClusterEngine({
    foobar: true,
    version: '1.19.1' // or AuroraVersion.1_19_1,
    // ...
  }),
});
```

What is a real world example of `foobar` in this case? I assumed these (i.e., `s3ImportRole`) are an example of `foobar`.

Reply:
`engineMode` is an example of `foobar`.

Reply:
> As an example, the current (i.e. before this PR) experience allows customers to specify `s3ImportRole` for even non-Aurora databases which will result in a validation error.

Can you show me where that is? Because I actually don't think that's true.

Reply:
> I was hoping that we will be solving the problem where all properties are available on instances/clusters irrespective of whether they are supported or not.
>
> The best experience would be to present the user only with the options that the selected database engine supports.
> 
> Did you ever intend for this to be fixed here, or was this not your intention?

I thought we talked about this topic when we discussed how this API should look like, and we decided that the amount of axes an engine can vary on (family, version, mode, etc.) is too big to make a separate class for each variant. So, each engine can add runtime validation to make sure the properties of the Cluster/Instance is being used for are correct (and, in your example, error out if it doesn't support S3 imports and/or exports).